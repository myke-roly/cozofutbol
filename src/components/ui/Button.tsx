import React, { memo, useCallback, useMemo, useRef } from 'react'
import {
  ButtonProps as RNButtonProps,
  StyleSheet,
  Text,
  Pressable,
  ViewProps,
  Animated,
} from 'react-native'
import { Color } from '../../constants'

export type ButtonProps = {
  style?: ViewProps['style']
  variant?: 'primary' | 'outline' | 'text'
} & RNButtonProps

const Button = ({
  title,
  onPress,
  style,
  variant,
  disabled,
  ...props
}: ButtonProps) => {
  const animatedOpacity = useRef(new Animated.Value(1)).current

  const isOutline = variant === 'outline'
  const isText = variant === 'text'

  const buttonStyle = useMemo(
    () =>
      StyleSheet.flatten([
        styles.button,
        isOutline && outlineStyles.button,
        isText && textStyles.button,
        style,
      ]),
    [isOutline, isText, style],
  )

  const textStyle = useMemo(
    () =>
      StyleSheet.flatten([
        styles.text,
        isOutline && outlineStyles.text,
        isText && textStyles.text,
      ]),
    [isOutline, isText],
  )

  const handleOnPressIn = useCallback(() => {
    Animated.timing(animatedOpacity, {
      toValue: 0.98,
      duration: 100,
      useNativeDriver: true,
    }).start()
  }, [animatedOpacity, disabled])

  const handleOnPressOut = useCallback(() => {
    Animated.timing(animatedOpacity, {
      toValue: 1,
      duration: 100,
      useNativeDriver: true,
    }).start()
  }, [animatedOpacity, disabled])

  const animatedStyles = useMemo(
    () => ({
      opacity: animatedOpacity,
      transform: [
        {
          scale: animatedOpacity,
        },
      ],
    }),
    [animatedOpacity],
  )

  console.log('Render button with title: ', title)

  return (
    <Pressable
      {...props}
      disabled={disabled}
      onPress={onPress}
      onPressIn={handleOnPressIn}
      onPressOut={handleOnPressOut}
      testID="btn-presable-view">
      <Animated.View style={[buttonStyle, animatedStyles]}>
        <Text style={textStyle}>{title}</Text>
      </Animated.View>
    </Pressable>
  )
}

Button.name = 'Button'

const styles = StyleSheet.create({
  button: {
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 8,
    borderWidth: 1,
    backgroundColor: Color.BLACK,
    paddingVertical: 16,
    paddingHorizontal: 8,
    shadowColor: Color.BLACK,
    shadowOffset: {
      width: 2,
      height: 2,
    },
    shadowOpacity: 1,
    elevation: 5,
  },
  text: {
    color: Color.WHITE,
    fontSize: 16,
  },
})

const outlineStyles = StyleSheet.create({
  button: {
    backgroundColor: 'transparent',
    borderColor: Color.BLACK,
    elevation: 0,
  },
  text: {
    color: Color.BLACK,
  },
})

const textStyles = StyleSheet.create({
  button: {
    backgroundColor: 'transparent',
    borderWidth: 0,
    elevation: 0,
  },
  text: {
    color: Color.BLACK,
  },
})

export default memo(Button)
