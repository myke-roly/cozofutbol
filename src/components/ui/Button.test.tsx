import renderer from 'react-test-renderer'
import Button from './Button'

describe('Button', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Button title="test-title" onPress={jest.fn()} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with disabled', () => {
    const tree = renderer
      .create(<Button title="test-title" onPress={jest.fn()} disabled />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with outline variant', () => {
    const tree = renderer
      .create(
        <Button title="test-title" onPress={jest.fn()} variant="outline" />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with text variant', () => {
    const tree = renderer
      .create(<Button title="test-title" onPress={jest.fn()} variant="text" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
