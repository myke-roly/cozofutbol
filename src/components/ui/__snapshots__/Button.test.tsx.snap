// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="btn-presable-view"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#232323",
        "borderRadius": 8,
        "borderWidth": 1,
        "elevation": 5,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 16,
        "shadowColor": "#232323",
        "shadowOffset": {
          "height": 2,
          "width": 2,
        },
        "shadowOpacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#fafafa",
          "fontSize": 16,
        }
      }
    >
      test-title
    </Text>
  </View>
</View>
`;

exports[`Button renders correctly with disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="btn-presable-view"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#232323",
        "borderRadius": 8,
        "borderWidth": 1,
        "elevation": 5,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 16,
        "shadowColor": "#232323",
        "shadowOffset": {
          "height": 2,
          "width": 2,
        },
        "shadowOpacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#fafafa",
          "fontSize": 16,
        }
      }
    >
      test-title
    </Text>
  </View>
</View>
`;

exports[`Button renders correctly with outline variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="btn-presable-view"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#232323",
        "borderRadius": 8,
        "borderWidth": 1,
        "elevation": 0,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 16,
        "shadowColor": "#232323",
        "shadowOffset": {
          "height": 2,
          "width": 2,
        },
        "shadowOpacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#232323",
          "fontSize": 16,
        }
      }
    >
      test-title
    </Text>
  </View>
</View>
`;

exports[`Button renders correctly with text variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="btn-presable-view"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderRadius": 8,
        "borderWidth": 0,
        "elevation": 0,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 16,
        "shadowColor": "#232323",
        "shadowOffset": {
          "height": 2,
          "width": 2,
        },
        "shadowOpacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#232323",
          "fontSize": 16,
        }
      }
    >
      test-title
    </Text>
  </View>
</View>
`;
